[deepSleep] Improve deepSleep task handling

clumsy-stefan opened the pull request:
added SensorSendAll() to ensure each task is run at least once after WiFi is initialized before going to deep sleep again

clumsy-stefan committed:
[deepSleep] Improve deepSleep task handling

added SensorSendAll() to ensure each task is run at least once after WiFi is initialized before going to deep sleep again

TD-er commented:
Not really clear enough in my head right now, but I was wondering if it will also send something when no data was read.
So is it now just sending the values stored in the UserVar?

clumsy-stefan commented:
I assume that depends on the plugin. Eg. the Sysinfo Plugin sets the UserVar within the `PLUGIN_READ` function, so it will send what is read there. I think for others (like ADC) similar, if there are no 10persec oder similar functions and the UserVars are set within `PLUGIN_READ` it will only be set/updated there.

TD-er commented:
The values will also be restored from the last session (saved in RTC)
So any plugin will have a value at boot that may not represent the most current environment state.
If the awake time is chosen so close to reading time (or WiFi connection may be fluctuating), then it may or may not take a reading. But with this change it will send out the last known value.
So issues caused by not actually reading may remain unnoticed for a long time

clumsy-stefan commented:
hmm... so there is also a dependency between the task interval and the sleep (and awake) time? meaning, that it's not guaranteed that a new reading is tried to be taken everytime when it awakes?
Eg. if the plugin only uses the `PLUGIN_READ` function with an Interval of say 900sec. and the device wakes every 300sec. it only takes a new reading every third time it awakes (roughly)?

TD-er commented:
Nope, the interval timer is reset at boot. Waking up is a special case of booting.
Only the "output values" are stored in the UserVar array and that array is stored in RTC.

clumsy-stefan commented:
> But with this change it will send out the last known value.
So issues caused by not actually reading may remain unnoticed for a long time

Why this? if the plugin can't read a value it will/shuold drop an error (at least in the log) and invalidate the UserVar. So you would actually catch it.

TD-er commented:
A lot of plugins do not update the UserVar value when there is no valid value read.
Others do update the UserVar value, but do not (yet) call the controller to send out data. For example a plugin that may need to read several values may only update one at a time.
By forcing the send to controller you may get intermediate results. For example an updated temperature but not yet updated humidity.
If you then compute the dew point, it may fluctuate a lot and some use cases make decisions based on the dew point. (e.g. to prevent mold grow)

clumsy-stefan commented:
I see, so that's also only a partial solution... seems to solve some of the issues, but not the underlying problem...

so still not sure how to tackle the issues, that some sensors/values are called "too early" in the whole process while making sure everything is read before going to sleep and give enough time to send it.....

clumsy-stefan commented:
I did a few more tests and had some issues with certain sensors (eg. dallas), and found that when I use your original commented out code on line 557:
```
schedule_all_task_device_timers(); // Disabled for now, since we are now using queues for controllers.
```
instead of what I proposed in this PR it seems to work more stable and reliable...

EDIT: however the `sleep_awake_time` seems to be essential and different for each unit depending on WiFi conenction/rate, CPU, server responsivness and others, one need to make sure it stays awake long enough to transmit all values before going to sleep again...

TD-er commented:
Please note that the `sleep_awake_time` is being reset at the moment the WiFi connection is successful.
So it does serve 2 purposes:
- WiFi timeout
- Process the tasks.

clumsy-stefan commented:
after 2 days of testing with a couplse of different depp-sleep nodes for me the best results werde when using the `schedule_all_task_device_timers()` functions at the time when WiFi connection is estabslihed and allow for enough time to be awake to get all data...

So I'd vote to include/activate this line again in future versions and change the sequence of initialisations (first controller plugins, then notifications, then the normal plugins).

I'll push it here later, so you can have a look, otherwise we can just delete this PR.

chemmex commented:
Is there currently any difference in behaviour between rules-initiated and webinterface-controlled deepsleep?

TD-er commented:
> Is there currently any difference in behaviour between rules-initiated and webinterface-controlled deepsleep?

Until a few months ago there was a difference, but now they should all use the same functions.

clumsy-stefan committed:
[deepSleep] Change Init-Sequence

- Change Init sequence: 1) Controllers, 2) Notifications, 3) Plugins to ensure controller queue's are initialized before first reading is taken
- use `schedule_all_task_device_timers()` after WiFi connection is established instead of `SensorSendAll()` (reverting changes from previous commit)

TD-er commented:
Is this PR still needed with my recent changes I merged this morning?

clumsy-stefan commented:
I think the only difference that remains is line 557 in `ESPEasy.ino`
```
schedule_all_task_device_timers(); // reenabled again to ensure one (more) run of tasks after WiFi connection when using deepSleep
```
which I need to ensure all tasks are called once (more) before deep sleep, especially the `PLUGIN_READ` functions. My deepSleep units won't send everything without this line.

So Unsure if this makes sense or not...

TD-er commented:
Can you move that change to a new issue (not a PR), because I don't know yet what effect it has on other use cases and maybe the true fix is to add a PLUGIN_PRE_SLEEP or something like we discussed.

clumsy-stefan commented:
opened a new issue as suggested, so from my point of view this PR can be deleted/closed and do fixes in a new PR as soon as it's more cleear on how it should be handled correctly. Until then I'll keep the changes I need (line above) locally.
